File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Java/src/fayelab/tdd/arithmetic/doublestack/negint/Expression.java
package fayelab.tdd.arithmetic.doublestack.negint;

public class Expression
{
    private Context ctx;
    private DoubleStackProcessor dsp;

    public Expression(String expr)
    {
        this(new Context(expr, 0));
    }

    Expression(Context ctx)
    {
        this.ctx = ctx;
        this.dsp = new DoubleStackProcessor();
    }
    
    public int eval()
    {
        StringBuffer digitBuffer = new StringBuffer();
        StringBuffer operatorBuffer = new StringBuffer();
        
        while(ctx.exprNotEnd())
        {
            char c = ctx.nextChar();
            if('(' == c)
            {
                //else operatorBuffer.length() == 0 when operator followed by left parenthesis
                processOperator(operatorBuffer);
                dsp.pushOperand(new Expression(ctx).eval());
            }
            else if(')' == c)
            {
                //else digitBuffer.length() == 0 when right parenthesis followed by right parenthesis
                processOperand(digitBuffer);
                return dsp.result();
            }
            else
            {
                if(isDigit(c))
                {
                    //else operatorBuffer.length() == 0 when operator followed by digit
                    processOperator(operatorBuffer);
                    digitBuffer.append(c);
                }
                else if(isMinusLeadingExpr(c))
                {
                    digitBuffer.append(c);
                }
                else
                {
                    //else digitBuffer.length() == 0 when right parenthesis followed by an operator
                    processOperand(digitBuffer);
                    operatorBuffer.append(c);
                }
            }
        }
        
        //else digitBuffer.length() == 0 when expr ends with right parenthesis
        processOperand(digitBuffer);
        return dsp.result();
    }
    
    private void processOperand(StringBuffer digitBuffer)
    {
        processOperElement(digitBuffer); 
    }

    private void processOperator(StringBuffer operatorBuffer)
    {
        processOperElement(operatorBuffer); 
    }
    
    private void processOperElement(StringBuffer buffer)
    {
        if(buffer.length() != 0)
        {
            dsp.process(buffer.toString());
            buffer.delete(0, buffer.length());
        }
    }

    private boolean isMinusLeadingExpr(char c)
    {
        return c == '-' && (ctx.isCurCharLeadingExpr() || '(' == ctx.prevChar());
    }


    private boolean isDigit(char c)
    {
        return '0' <= c && c <= '9';
    }
}

class Context
{
    private String expr;
    private int curIdx;
    
    Context(String expr, int curIdx)
    {
        this.expr = expr;
        this.curIdx = curIdx;
    }

    boolean exprNotEnd()
    {
        return curIdx < expr.length();
    }

    char nextChar()
    {
        return expr.charAt(curIdx++);
    }
    
    boolean isCurCharLeadingExpr()
    {
        return curIdx == 1;
    }
    
    public char prevChar()
    {
        return expr.charAt(curIdx - 2);
    }
}

File: TDD/TDD Coding Practice Arithmetic/Double Stack Algorithm/Code/Python/oo/multidigit/expression.py
from doublestackprocessor import DoubleStackProcessor

class Expression(object):
    def __init__(self, expr, context=None):
        if context == None:
            self.ctx = Context(expr, 0)
        else:
            self.ctx = context

        self.dsp = DoubleStackProcessor()

    def eval(self):
        digitbuffer = ''

        while self.ctx.expr_not_end():
            c = self.ctx.next_char()
            if '(' == c:
                self.dsp.push_operand(Expression(self.ctx.expr, self.ctx).eval())
            elif ')' == c:
                if len(digitbuffer) != 0: # else len(digitbuffer) == 0 when right parenthesis followed by right parenthesis
                    self.dsp.process(digitbuffer)

                return self.dsp.result()
            else:
                if c.isdigit():
                    digitbuffer += c
                else:
                    if len(digitbuffer) != 0: # else len(digitbuffer) == 0 when right parenthesis followed by an operator
                        self.dsp.process(digitbuffer)
                        digitbuffer = ''

                    self.dsp.process(c)

        if len(digitbuffer) != 0:
            self.dsp.process(digitbuffer) # else len(digitbuffer) == 0 when expr ends with right parenthesis

        return self.dsp.result()

class Context(object):
    def __init__(self, expr, curidx):
        self.expr = expr
        self.curidx = curidx

    def expr_not_end(self):
        return self.curidx < len(self.expr)

    def next_char(self):
        c = self.expr[self.curidx]
        self.curidx += 1
        return c